executor/indexer.py

```diff
@@ -0,0 +1,368 @@
+import logging
+from copy import deepcopy
+
+from collections import OrderedDict
+from pathlib import Path
+from typing import Optional, Dict, List, Union
+import os
+import faiss
+import numpy as np
+from bloom_filter2 import BloomFilter
+
+from jina import Document, Executor, DocumentArray, requests
+from jina.logging.logger import JinaLogger
+from .storage import StorageFactory
+
+
+class FaissIndexer(Executor):
+    """A vector similarity indexer for very large scale data using Faiss.
+
+    The documents can be stored using different storage backend (e.g., LMDB, SQLite, PostgresSQL),
+    while the vector embeddings are indexed in a FAISS Index.
+    """
+
+    def __init__(
+        self,
+        storage_backend: str = 'lmdb',
+        index_key: str = 'Flat',
+        metric: str = 'cosine',
+        trained_index_file: Optional[str] = None,
+        *args,
+        **kwargs,
+    ):
+        """
+        :param storage_backend: the storage backend, e.g., ``LMDB``, ``postgresql+psycopg2://``
+        :param index_key: create a new FAISS index of the specified type.
+                The type is determined from the given string following the conventions
+                of the original FAISS index factory.
+                    Recommended options:
+                    - "Flat" (default): Best accuracy (= exact). Becomes slow and RAM intense for > 1 Mio docs.
+                    - "HNSW": Graph-based heuristic. If not further specified,
+                        we use a RAM intense, but more accurate config:
+                        HNSW256, efConstruction=256 and efSearch=256
+                    - "IVFx,Flat": Inverted Index. Replace x with the number of centroids aka nlist.
+                        Rule of thumb: nlist = 10 * sqrt (num_docs) is a good starting point.
+                    For more details see:
+                    - Overview of indices https://github.com/facebookresearch/faiss/wiki/Faiss-indexes
+                    - Guideline for choosing an index https://github.com/facebookresearch/faiss/wiki/Guidelines-to-choose-an-index
+                    - FAISS Index factory https://github.com/facebookresearch/faiss/wiki/The-index-factory
+        :param trained_index_file: the index file dumped from a trained index, e.g., ``faiss.index``.
+        """
+        super().__init__(*args, **kwargs)
+        self.logger = JinaLogger(self.__class__.__name__)
+
+        self.index_key = index_key
+        self.metric = metric
+
+        workspace = Path(self.workspace)
+        storage_path = str(workspace / 'lmdb_docs')
+
+        self._metas = {'doc_ids': [], 'doc_id_to_offset': {}, 'delete_marks': []}
```

Comment:
You are using two dictionaries, `doc_ids` and `doc_id_to_offset`. The problem that you have here is that you have to manually make sure they are in sync. Consider using `bidict` package for that.

Reply:
thanks for your suggestion. `bidict` is perfect, I would like to use it.

Reply:
BTW, May I ask for your help to make this update (if you are available)? https://github.com/jina-ai/executor-faissindexer/issues/8

Reply:
Sure, I can work on it 👍